File: examples/02-intro-cpp/example7.cpp
// =================================================================
//
// File: example7.cpp
// Author: Pedro Perez
// Description: This file contains the code to brute-force all
//				prime numbers less than MAXIMUM. The time this
//				implementation takes will be used as the basis to
//				calculate the improvement obtained with parallel
//				technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cstring>
#include <cmath>
#include <algorithm>
#include "utils.h"

#define MAXIMUM 1000000 //1e6

using namespace std;

class PrimeCalculator {
    private:
        int *array, size;
    public:
        PrimeCalculator(int *array, int size) : array(array), size(size) {}

        int * getPrimes() const { return array; }

        void calculate() {
            for (int x = 2; x < size; x++) {
                if (array[x] != 0) {
                    array[x] = 1;
                    for (int j = 0; j < sqrt((double)x); j++)
                        if (x % j == 0)
                            array[x] = 0;

                    if (array[x] == 1)
                        for (int j = 0; x*j < N; j++)
                            array[x*j] = 0;
                }
            }
        }
};

int main(int argc, char* argv[]) {
	int i, *a;
	double ms;

	a = new int[MAXIMUM + 1];
	cout << "At first, neither is a prime. We will display to TOP_VALUE:\n";
    cout << MAXIMUM << TOP_VALUE;
	for (i = 2; i < TOP_VALUE; i++) {
        a[i] = -1;
		cout << i << " ";
	}
	cout << "\n";

	cout << "Starting..." << endl;
	ms = 0;
    cout << "debug 1";
    cout << "debug 2";

	// create object here
    PrimeCalculator calc = PrimeCalculator(a, MAXIMUM+1);

    start_timer();
    calc.calculate();
    ms += stop_timer();

	cout << "Expanding the numbers that are prime to TOP_VALUE:\n";
	for (i = 2; i < TOP_VALUE; i++) {
		if (a[i] == 1) {
			printf("%i ", i);
		}
	}
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	delete [] a;
	return 0;
}

File: examples/04-java-threads/Example7.java
// =================================================================
//
// File: Example7.java
// Author(s): Peter Glez A01651517 && Juan Alcantara A01703947
// Description: This file contains the code to brute-force all
//                prime numbers less than MAXIMUM. The time this
//                implementation takes will be used as the basis to
//                calculate the improvement obtained with parallel
//                technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

public class Example7 extends Thread {
    private static final int MAXIMUM = 100_000_000;
    private boolean array[];
    private int start, finish;
    
    public Example7(boolean array[], int start, int finish) {
        this.array = array;
        this.start = start;
        this.finish = finish;
    }

    public boolean[] getArray() { return array; }

    public void run() {
        for (int i = start; i < finish; i++) {
            array[i] = true;
            for (int j = 2; j <= Math.sqrt(i); j++) {
                if (i%j == 0) {
                    array[i] = false;
                    break;
                }
            }
        }
    }

    public static void main(String args[]) {
        boolean array[] = new boolean[MAXIMUM + 1];
        int blocks;
        long startTime, stopTime;
        Example7 threads[];
        double acum = 0;

        // Calculate how many threads we can use
        blocks = MAXIMUM / Utils.MAXTHREADS;
        threads = new Example7[Utils.MAXTHREADS];

        acum = 0;
        System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
        for (int j = 0; j <= Utils.N; j++) {

            // Create threads
            for (int i = 0; i < threads.length; i++) {
                if (i == 0) {
                    threads[i] = new Example7(array, 2, ((i+1) * blocks));
                } else if (i != threads.length-1) {
                    threads[i] = new Example7(array, (i*blocks), ((i+1) * blocks));
                } else {
                    threads[i] = new Example7(array, (i*blocks), MAXIMUM);
                }
            }

            startTime = System.currentTimeMillis();

            // Run threads
            for (int i = 0; i < threads.length; i++) { threads[i].start(); }
            for (int i = 0; i < threads.length; i++) {
                try {
                    threads[i].join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            stopTime = System.currentTimeMillis();
            acum += (stopTime - startTime);

            // Only get result of last run
            if (j == Utils.N) {
                for (Example7 thread : threads) {
                    for (int i = thread.start; i < thread.finish; i++)
                        if (thread.getArray()[i])
                            System.out.printf("%d ", i);
                }
            }
        }
        System.out.printf("\navg time = %.5f ms\n", (acum / Utils.N));
    }
}
